File: FBSimulatorControl/Commands/FBSimulatorLogCommands.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBSimulatorLogCommands.h"

#import <CoreSimulator/SimDevice.h>
#import <CoreSimulator/SimRuntime.h>

#import "FBSimulator+Private.h"
#import "FBSimulatorError.h"

@interface FBSimulatorLogCommands ()

@property (nonatomic, weak, readonly) FBSimulator *simulator;

@end

@implementation FBSimulatorLogCommands

#pragma mark Initializers

+ (instancetype)commandsWithTarget:(FBSimulator *)target
{
  return [[self alloc] initWithSimulator:target];
}

- (instancetype)initWithSimulator:(FBSimulator *)simulator
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _simulator = simulator;

  return self;
}

#pragma mark Public

- (FBFuture<id<FBLogOperation>> *)tailLog:(NSArray<NSString *> *)arguments consumer:(id<FBDataConsumer>)consumer
{
  return [[self
    startLogCommand:[FBProcessLogOperation osLogArgumentsInsertStreamIfNeeded:arguments] consumer:consumer]
    onQueue:self.simulator.workQueue map:^(FBProcess *process) {
      return [[FBProcessLogOperation alloc] initWithProcess:process consumer:consumer queue:self.simulator.asyncQueue];
    }];
}

#pragma mark Private

- (FBFuture<FBProcess *> *)startLogCommand:(NSArray<NSString *> *)arguments consumer:(id<FBDataConsumer>)consumer
{
  NSError *error = nil;
  NSString *launchPath = [self logExecutablePathWithError:&error];
  if (!launchPath) {
    return [FBSimulatorError failFutureWithError:error];
  }
  FBProcessIO *processIO = [[FBProcessIO alloc]
    initWithStdIn:nil
    stdOut:[FBProcessOutput outputForDataConsumer:consumer]
    stdErr:nil];

  FBProcessSpawnConfiguration *configuration = [[FBProcessSpawnConfiguration alloc]
    initWithLaunchPath:launchPath
    arguments:arguments
    environment:@{}
    io:processIO
    mode:FBProcessSpawnModeDefault];

  return [self.simulator launchProcess:configuration];
}

- (NSString *)logExecutablePathWithError:(NSError **)error
{
  NSString *path = [[[self.simulator.device.runtime.root
    stringByAppendingPathComponent:@"usr"]
    stringByAppendingPathComponent:@"bin"]
    stringByAppendingPathComponent:@"log"];
  FBBinaryDescriptor *binary = [FBBinaryDescriptor binaryWithPath:path error:error];
  if (!binary) {
    return nil;
  }
  return binary.path;
}

@end

File: FBSimulatorControl/Commands/FBSimulatorLogCommands.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@preconcurrency import CoreSimulator
import FBControlCore
import Foundation

/// The way log invocation fails without a runtime root, as data rather than an assembled string.
public enum FBSimulatorLogError: Error, LocalizedError {
  case runtimeRootUnavailable

  public var errorDescription: String? {
    switch self {
    case .runtimeRootUnavailable:
      return "Could not obtain runtime root for simulator"
    }
  }
}

public final class FBSimulatorLogCommands: NSObject {

  // MARK: - Properties

  private weak var simulator: FBSimulator?

  // MARK: - Initializers

  public class func commands(with simulator: FBSimulator) -> FBSimulatorLogCommands {
    FBSimulatorLogCommands(simulator: simulator)
  }

  private init(simulator: FBSimulator) {
    self.simulator = simulator
    super.init()
  }

  // MARK: - Private

  fileprivate func tailLogAsync(arguments: [String], consumer: any FBDataConsumer) async throws -> any LogOperation {
    guard let simulator = self.simulator else {
      throw FBWeakTargetError.simulator
    }
    let launchPath = try logExecutablePath()
    let streamArguments = FBProcessLogOperation.osLogArgumentsInsertStreamIfNeeded(arguments)
    let processIO = FBProcessIO<AnyObject, AnyObject, AnyObject>(
      stdIn: nil,
      stdOut: FBProcessOutput<AnyObject>(for: consumer),
      stdErr: nil
    )
    let configuration = FBProcessSpawnConfiguration(
      launchPath: launchPath,
      arguments: streamArguments,
      environment: [:],
      io: processIO,
      mode: .default
    )
    let process = try await simulator.launchProcess(configuration)
    return FBProcessLogOperation(process: process, consumer: consumer, queue: simulator.asyncQueue)
  }

  private func logExecutablePath() throws -> String {
    guard let simulator = self.simulator else {
      throw FBWeakTargetError.simulator
    }
    guard let root = simulator.device.runtime.root else {
      throw FBSimulatorLogError.runtimeRootUnavailable
    }
    let path =
      (((root as NSString)
      .appendingPathComponent("usr") as NSString)
      .appendingPathComponent("bin") as NSString)
      .appendingPathComponent("log")
    let binary = try FBBinaryDescriptor.binary(withPath: path)
    return binary.path
  }
}

// MARK: - FBSimulator+LogCommands

extension FBSimulator: LogCommands {

  public func tailLog(arguments: [String], consumer: any FBDataConsumer) async throws -> any LogOperation {
    return try await logCommands().tailLogAsync(arguments: arguments, consumer: consumer)
  }
}
